File: PHP/1071. Greatest Common Divisor of Strings.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    public function gcdOfStrings(string $str1, string $str2): string
    {
        if ($str1.$str2 !== $str2.$str1) {
            return '';
        }

        return substr($str1, 0, $this->gcd(strlen($str1), strlen($str2)));
    }

    /** Euclidean Algorithm */
    public function gcd(int $former, int $latter): int
    {
        if ($latter === 0) {
            return $former;
        }

        return $this->gcd($latter, $former % $latter);
    }
}

File: Python/1071. Greatest Common Divisor of Strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if str1 + str2 != str2 + str1:
            return ''

        length = self.gcd(len(str1), len(str2))
        return str1[:length]

    def gcd(self, a, b):
        return a if b == 0 else self.gcd(b, a % b)
